NewApp/ngsa-csharp/Ngsa.Middleware/NgsaLog.cs

```diff
@@ -0,0 +1,174 @@
+        public HttpContext Context { get; set; } = null;
+        public Dictionary<string, string> Data { get; } = new Dictionary<string, string>();
+
+        public NgsaLog GetLogger(string method, HttpContext context = null)
+        {
+            NgsaLog logger = new NgsaLog
+            {
+                Name = Name,
+                ErrorMessage = ErrorMessage,
+                NotFoundError = NotFoundError,
+                LogLevel = LogLevel,
+
+                Method = method,
+                Context = context,
+            };
+
+            // get the next key
+            while (Loggers.ContainsKey(counter))
+            {
+                if (counter == int.MaxValue)
+                {
+                    counter = 0;
+                }
+
+                counter++;
+            }
+
+            // todo - reuse loggers
+            // use iDisposable?
+            //Loggers.Add(counter, logger);
+
+            return logger;
+        }
+
+        public void LogInformation(string message)
+        {
+            const LogLevel logLevel = LogLevel.Information;
+
+            if (LogLevel <= logLevel)
+            {
+                Dictionary<string, object> d = GetDictionary(message, logLevel);
+
+                Console.ForegroundColor = ConsoleColor.Green;
+                Console.WriteLine(JsonSerializer.Serialize(d, Options));
+                Console.ResetColor();
+            }
+        }
+
+        public void LogWarning(string message)
+        {
+            const LogLevel logLevel = LogLevel.Warning;
+
+            if (LogLevel > logLevel)
+            {
+                return;
+            }
+
+            Dictionary<string, object> d = GetDictionary(message, logLevel);
+
+            Console.ForegroundColor = ConsoleColor.Yellow;
+            Console.WriteLine(JsonSerializer.Serialize(d, Options));
+            Console.ResetColor();
+        }
+
+        public void LogError(string message, Exception ex = null)
+        {
+            const LogLevel logLevel = LogLevel.Error;
+
+            if (LogLevel > logLevel)
+            {
+                return;
+            }
+
+            Dictionary<string, object> d = GetDictionary(message, logLevel);
+
+            if (ex != null)
+            {
+                d.Add("ExceptionType", ex.GetType());
+                d.Add("ExceptionMessage", ex.Message);
+
+                //d.Add("exceptionText", ex.ToString());
+            }
+            else if (Exception != null)
+            {
+                d.Add("ExceptionType", Exception.GetType().FullName);
+                d.Add("ExceptionMessage", Exception.Message);
+
+                //d.Add("exceptionText", Exception.ToString());
+            }
+
+            // display the error
+            Console.ForegroundColor = ConsoleColor.Red;
+            Console.Error.WriteLine(JsonSerializer.Serialize(d, Options));
+            Console.ResetColor();
+        }
```

Comment:
The 3 Console lines repeat.  Could move into a common WriteMessage(color, dict) method.

Reply:
I thought about that, but only 2 of them repeat as one uses console.error. Didn't seem like enough reuse to justify.